src/common/pool_map.c

```diff
@@ -2113,6 +2115,427 @@ pool_map_update_failed_cnt(struct pool_map *map)
+					      ver.pf_end_ver);
+			return 0;
+		}
+	}
+
+	D_ASSERT(fnode->pf_ver_nr <= fnode->pf_ver_total);
+	if (fnode->pf_ver_nr == fnode->pf_ver_total) {
+		nr = fnode->pf_ver_nr + PMAP_FAIL_INLINE_NR;
+		D_ALLOC_ARRAY(fvers, nr);
+		if (fvers == NULL)
+			return -DER_NOMEM;
+
+		memcpy(fvers, fnode->pf_vers,
+		       sizeof(*fvers) * fnode->pf_ver_nr);
+		if (fnode->pf_vers != fnode->pf_ver_inline)
+			D_FREE(fnode->pf_vers);
+		fnode->pf_vers = fvers;
+		fnode->pf_ver_total = nr;
+	}
+	D_ASSERT(fnode->pf_ver_nr < fnode->pf_ver_total);
+
+	fnode->pf_vers[fnode->pf_ver_nr] = ver;
+	fnode->pf_ver_nr++;
+	return 0;
+}
+
+static int
+pmap_node_check(struct pool_domain *node_dom, struct pmap_fail_stat *fstat)
+{
+	struct pmap_fail_node	*fnode = NULL;
+	struct pool_target	*tgt;
+	struct pool_component	*comp;
+	int			 i;
+	int			 rc = 0;
+
+	for (i = 0; i < node_dom->do_target_nr; ++i) {
+		tgt = &node_dom->do_targets[i];
+		comp = &tgt->ta_comp;
+		if (!pmap_comp_failed(comp) ||
+		    pmap_comp_execlude_out_earlier(comp, fstat->pf_last_ver))
+			continue;
+		if (fnode == NULL) {
+			fnode = pmap_fail_node_get(fstat);
+			if (fnode == NULL)
+				return -DER_NOMEM;
+		}
+
+		rc = pmap_fail_node_add_tgt(fstat, fnode, comp);
+		if (rc)
+			return rc;
+	}
+
+	if (fnode == NULL || fnode->pf_ver_nr == 0)
+		return 0;
+
+	fnode->pf_co_rank = node_dom->do_comp.co_rank;
+	daos_array_sort(fnode->pf_vers, fnode->pf_ver_nr, false,
+			&pmap_fver_sort_ops);
+	pmap_fail_ver_merge(fnode);
+
+	if (fnode->pf_down)
+		fstat->pf_down_nr++;
+	if (fnode->pf_new_fail)
+		fstat->pf_newfail_nr++;
+	if ((fstat->pf_down_nr > fstat->pf_rf) && (fstat->pf_newfail_nr > 0)) {
+		rc = -DER_RF;
+		D_ERROR("RF broken, found %d DOWN node, newly fail %d, rf %d, "
+			DF_RC"\n", fstat->pf_down_nr, fstat->pf_newfail_nr,
+			fstat->pf_rf, DP_RC(rc));
+	}
+
+	return rc;
+}
+
+static int
+pmap_fail_ver_overlap(struct pmap_fail_stat *fstat, struct pmap_fail_ver *fver)
+{
+	struct pmap_fail_node	*fnode;
+	struct pmap_fail_ver	*tmp;
+	int			 i, j;
+	int			 nr;
+	bool			 ovl;
+
+	for (i = 0, nr = 0; i < fstat->pf_node_nr; i++) {
+		ovl = false;
+		fnode = &fstat->pf_nodes[i];
+		for (j = 0; j < fnode->pf_ver_nr; j++) {
+			tmp = &fnode->pf_vers[j];
+			if (fver_overlap(tmp, fver)) {
+				ovl = true;
+				break;
+			}
+		}
+		if (ovl)
+			nr++;
+	}
+
+	return nr;
+}
+
+static int
+pmap_fail_stat_check(struct pmap_fail_stat *fstat)
+{
+	struct pmap_fail_node	*fnode;
+	struct pmap_fail_ver	*fver;
+	uint32_t		 rf = fstat->pf_rf;
+	int			 max_fail_nr = 0;
+	int			 fail_nr;
+	uint32_t		 i, j;
+	int			 rc = 0;
+
+	/* First check some easier cases, it should cover most common cases */
+	if ((fstat->pf_node_nr <= rf) || (fstat->pf_newfail_nr == 0))
+		return 0;
+	if (fstat->pf_node_nr == 1) {
+		if (rf >= 1)
+			return 0;
+		rc = -DER_RF;
+	} else if (fstat->pf_down_nr > rf) {
+		rc = -DER_RF;
+	}
+	if (rc)
+		goto fail;
+
+	/* Bad cornel case, need to check #max-concurrent-failures by checking
```

Comment:
s/cornel/corner

Reply:
I'll fix it in following patch, for when this patch need refresh later. Thanks.